### PR TITLE
Use dune 2.8 to build the project

### PR DESCRIPTION
**Pull request by psafont**
Changing the build system to use dune allows out-of-the-box compatibility with ctypes-foreign 0.18 and greater. Dune 2.8 is needed in order to keep several libraries private, previously they needed to be public in order to be used by public libraries.

This change also generates 3 new libraries public, unfortunately. They are not useful to the users and are a byproduct on not being able to have private libraries inside public ones they are unix_errno.bindings, unix_errno.types and unix_errno.types.detected.

I need to investigate whether the change is making dlm to fail to link when using ctypes 0.18.0 while it works when using lower versions of ctypes.

Comments, suggestions and pointers from anybody reading the PR are very much welcome, the porting process was quite complicated to me and might have missed several issues.

Updated to dune 2.8 to keep most libraries private:

```
$ opam show unix-errno --list-files
/home/paus/.local/share/opam/toolstack-local/bin/errno-map
/home/paus/.local/share/opam/toolstack-local/doc/unix-errno
/home/paus/.local/share/opam/toolstack-local/doc/unix-errno/CHANGES
/home/paus/.local/share/opam/toolstack-local/doc/unix-errno/README.md
/home/paus/.local/share/opam/toolstack-local/lib/stublibs/dllunix_errno_stubs.so
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/META
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/.public_cmi/unix_errno_bindings.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/.public_cmi/unix_errno_bindings.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/errno_bindings.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/errno_bindings.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/errno_bindings.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/errno_bindings.cmxs
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/unix_errno_bindings.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_bindings/unix_errno_bindings.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/.public_cmi/unix_errno_generated.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/.public_cmi/unix_errno_generated.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/.public_cmi/unix_errno_types.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/.public_cmi/unix_errno_types.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/errno_types.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/errno_types.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/errno_types.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/errno_types.cmxs
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/unix_errno_generated.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/unix_errno_generated.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/unix_errno_types.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types/unix_errno_types.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/.public_cmi/unix_errno_types_detected.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/.public_cmi/unix_errno_types_detected.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/errno_types_detected.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/errno_types_detected.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/errno_types_detected.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/errno_types_detected.cmxs
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/unix_errno_types_detected.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/__private__/errno_types_detected/unix_errno_types_detected.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/dune-package
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmti
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.cmxs
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/errno.mli
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmti
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.cmxs
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_host.mli
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_linux_4_4_6.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_linux_4_4_6.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_linux_4_4_6.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_linux_4_4_6.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_osx_10_11_4.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_osx_10_11_4.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_osx_10_11_4.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/host/errno_map_osx_10_11_4.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/opam
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.cmi
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.cmt
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.cmti
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.cmx
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.ml
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/errno_unix.mli
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/libunix_errno_stubs.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/unix_errno.a
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/unix_errno.cma
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/unix_errno.cmxa
/home/paus/.local/share/opam/toolstack-local/lib/unix-errno/unix/unix_errno.cmxs
```

compared to the previous version:

```
$ opam show unix-errno --list-files
/home/paus/.local/share/opam/toolstack/bin/errno-map
/home/paus/.local/share/opam/toolstack/bin/errno-srcgen
/home/paus/.local/share/opam/toolstack/lib/stublibs/dllunix_errno_stubs.so
/home/paus/.local/share/opam/toolstack/lib/stublibs/dllunix_errno_stubs.so.owner
/home/paus/.local/share/opam/toolstack/lib/unix-errno
/home/paus/.local/share/opam/toolstack/lib/unix-errno/META
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno.cmi
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno.cmti
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno.cmx
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno.mli
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_.a
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_.cma
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_.cmxa
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_host.cmi
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_host.cmti
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_host.cmx
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_host.mli
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_unix.cmi
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_unix.cmti
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_unix.cmx
/home/paus/.local/share/opam/toolstack/lib/unix-errno/errno_unix.mli
/home/paus/.local/share/opam/toolstack/lib/unix-errno/libunix_errno_stubs.a
/home/paus/.local/share/opam/toolstack/lib/unix-errno/unix_errno.a
/home/paus/.local/share/opam/toolstack/lib/unix-errno/unix_errno.cma
/home/paus/.local/share/opam/toolstack/lib/unix-errno/unix_errno.cmxa
```

CI status can be found in https://github.com/psafont/ocaml-unix-errno/actions/runs/1136117659

### CODE REVIEW

**Comment by psafont** [comment]
I need to change the metadata for opam as David is no longer maintaining the package

**Review by lindig — APPROVED** [review]

File: unix-errno.opam [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+description: """\
+unix-errno can be used with or without ctypes and OCaml's Unix
+module. Without ctypes and Unix, the basic types and functions are
+provided as well as Errno_host containing errno maps for popular
+operating systems. The errno-srcgen tool for generating OCaml source
+representing Errno.Host.t values will also be built. With ctypes and
+Unix, you'll also receive the errno-map tool for outputting the current
+host's errno map and the Errno_unix module containing an errno global
+variable checking function and Unix.error type converters."""
+maintainer: "sheets@alum.mit.edu"
```
Comment (edwintorok) [inline unix-errno.opam]: probably needs changing

**Review by edwintorok — COMMENTED** [review]
Would be good to setup the CI too, to check we haven't missed anything (e.g. `git add`-ing a file)

**Comment by psafont** [comment]
Tests pass both on linux and macos: https://github.com/psafont/ocaml-unix-errno/actions/runs/1136117659

**Review by lippirk — APPROVED** [review]

**Pull request merged**